catalog/harvest/catalog-harvest-metacardadapter/src/main/java/org/codice/ddf/catalog/harvest/adapter/HarvestedResourceTransformer.java

```diff
@@ -0,0 +1,189 @@
+  public Metacard transformHarvestedResource(HarvestedResource resource, String metacardId) {
+    return transform(resource, metacardId);
+  }
+
+  private Metacard transform(HarvestedResource harvestedResource, String metacardId) {
+    MimeType resourceMimeType = harvestedResource.getMimeType();
+
+    Metacard metacard = null;
+
+    try (TemporaryFileBackedOutputStream tfbos = new TemporaryFileBackedOutputStream()) {
+      IOUtils.copy(harvestedResource.getInputStream(), tfbos);
+      if (resourceMimeType == null) {
+        resourceMimeType =
+            guessMimeTypeFor(
+                tfbos.asByteSource().openStream(),
+                FilenameUtils.getExtension(harvestedResource.getName()));
+      }
+
+      metacard =
+          doTransform(
+              resourceMimeType, tfbos.asByteSource().openStream(), harvestedResource, metacardId);
+
+      if (metacard != null) {
+        enrichMetacard(metacard, harvestedResource);
+        return metacard;
+      }
+    } catch (IOException e) {
+      LOGGER.debug("Failed to open TFBOS for harvested resource's input stream.", e);
+    }
+    return metacard;
+  }
+
+  private void enrichMetacard(Metacard metacard, HarvestedResource harvestedResource) {
+    writeMetacardAttribute(metacard, Core.TITLE, harvestedResource.getName());
+    writeMetacardAttribute(
+        metacard, Core.RESOURCE_SIZE, Long.toString(harvestedResource.getSize()));
+    writeMetacardAttribute(metacard, Core.RESOURCE_URI, harvestedResource.getUri().toASCIIString());
+  }
+
+  private Metacard doTransform(
+      MimeType mimeType, InputStream is, Resource resource, String metacardId) {
+    List<InputTransformer> transformerCandidates =
+        mimeTypeToTransformerMapper.findMatches(InputTransformer.class, mimeType);
+
+    for (InputTransformer inputTransformer : transformerCandidates) {
+      Optional<Metacard> metacardOptional =
+          doTransform(inputTransformer, metacardId, is, resource.getName());
+
+      if (metacardOptional.isPresent()) {
+        return metacardOptional.get();
+      }
+    }
+    return null;
+  }
+
+  private Optional<Metacard> doTransform(
+      InputTransformer inputTransformer, String metacardId, InputStream is, String resourceName) {
+    try {
+      if (StringUtils.isNotEmpty(metacardId)) {
+        return Optional.of(inputTransformer.transform(is, metacardId));
+      }
+
+      return Optional.of(inputTransformer.transform(is));
+    } catch (IOException e) {
+      LOGGER.debug("Failed to retrieve inputStream for resource [{}].", resourceName, e);
+    } catch (CatalogTransformerException e) {
+      LOGGER.debug(
+          "Failed to transform resource [{}] with [{}] transformer. Trying next one.",
+          resourceName,
+          inputTransformer);
+    }
+
+    return Optional.empty();
+  }
+
+  private MimeType guessMimeTypeFor(InputStream is, String fileExt) {
+    try (TemporaryFileBackedOutputStream tfbos = new TemporaryFileBackedOutputStream()) {
```

Comment:
❓ Do we really need to create another `TemporaryFileBackedOutputStream` here and make a copy again? We've already done that before calling this method (L86).